src/ocean/io/select/EpollSelectDispatcher.d

```diff
@@ -382,13 +387,27 @@ public class EpollSelectDispatcher : IEpollSelectDispatcherInfo
 
      **************************************************************************/
 
-    public int unregister ( ISelectClient client )
+    public int unregister ( ISelectClient client,
+        bool remove_from_selected_set = false )
     {
         try if (client.is_registered)
         {
             scope (success)
             {
                 this.registered_clients -= client;
+
+                if (remove_from_selected_set)
+                {
+                    scope match_client_ptr = (Const!(epoll_event_t) entry) {
+                        return entry.data.ptr == cast(void*)client;
+                    };
+
+                    auto index = this.selected_set.findIf(match_client_ptr);
+                    if (index < this.selected_set.length)
+                    {
+                        this.selected_set[index] = epoll_event_t.init;
```

Comment:
Maybe you could remove the array element i.e.
```d
this.selected_set[index] = this.selected_set[$ - 1];
this.selected_set = this.selected_set[0 .. $ - 1];
```
?
Also, just `foreach (selected_client; this.selected_set)` is probably less code (a matter of taste, though).

Reply:
Removing the array element is problematic because of two things:

1. When this is running, there's a good chance that `SelectedKeysHandler` is already doing `foreach (key; selected_keys)`, so we need to be very careful about changing the array while iterating.

2. How the things are setup at the moment, `EpollSelectDispatcher` has the array that both `EpollSelectDispatcher` and `SelectedKeysHandler` slice. Changing the `this.selected_set` slice will not reflect on the slices in `SelectedKeysHandler`, so we need to add the method to the `SelectedKeysHandler` to remove this array from it's slice, but also this will have to change the array stored here.

Since the only penalty here is checking with the null, and which would always pass, except for very obscure cases where the `ISelectClient` instances have the scoped lifetime, I don't think this is worth it.

Reply:
> Maybe you could remove the array element

As discussed verbally, this would interfere with the iteration in `SelectedKeysHandler.opCall` so never mind.